[Model Monitoring] Add TimescaleDB database auto-creation

alxtkr77 opened the pull request:
TITLE: [Model Monitoring] Add TimescaleDB database auto-creation

## Summary
Add automatic database creation for TimescaleDB connector. When `auto_create_database` is enabled (default), the connector generates a database name using `system_id` and creates it if it does not exist.


## Changes Made
- Add `tsdb.auto_create_database` config option to `mlrun/config.py` (default: True)
- Add `admin_dsn()` method to `DatastoreProfilePostgreSQL` for connecting to postgres maintenance database
- Add `dsn(database=...)` parameter to override database in DSN
- Add `_determine_database_name()` to `TimescaleDBConnector`
- Add `_create_db_if_not_exists()` to `TimescaleDBOperationsManager`
- Add unit tests for database auto-creation functionality (4 tests)
- Add unit tests for `admin_dsn()` and `dsn()` methods (2 tests)

## Testing
- 107/107 TimescaleDB tests pass
- 5/5 PostgreSQL datastore profile tests pass
- Tested database creation, idempotency, and error handling

## Checklist
- [x] Code follows project style guidelines
- [x] Tests added for new functionality
- [x] All tests pass locally
- [x] Documentation updated (internal docstrings)

## Reference
- Jira: [ML-11662](https://iguazio.atlassian.net/browse/ML-11662)


[ML-11662]: https://iguazio.atlassian.net/browse/ML-11662?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
[Model Monitoring] Add TimescaleDB database auto-creation (ML-11662)

Add automatic database creation for TimescaleDB connector. When
`auto_create_database` is enabled (default), the connector generates
a database name using `system_id` and creates it if it doesn't exist.

Changes:
- Add `tsdb.auto_create_database` config option (default: True)
- Add `admin_dsn()` method to `DatastoreProfilePostgreSQL`
- Add `dsn(database=...)` parameter override
- Add `_determine_database_name()` to `TimescaleDBConnector`
- Add `_create_db_if_not_exists()` to `TimescaleDBOperationsManager`
- Add unit tests for database auto-creation functionality

gtopper commented on mlrun/datastore/datastore_profile.py:
```diff
@@ -500,18 +500,37 @@ class DatastoreProfilePostgreSQL(DatastoreProfile):
     database: str = pydantic.v1.Field(
         default="postgres"
-    )  # the default maintenance database
+    )  # Default to postgres maintenance database
```
```suggestion
    database: str = "postgres"
```
Seems like unnecessary wrapper and a comment that misinterprets the unnecessary wrapper?

alxtkr77 replied:
Agreed, simplified.

gtopper commented on mlrun/model_monitoring/db/tsdb/timescaledb/timescaledb_connector.py:
```diff
@@ -118,6 +148,34 @@ def __init__(
+
+        if not auto_create:
+            # Use database from profile as-is
+            return profile.database or "postgres"
```
`profile.database` is already `"postgres"` though, isn't it? I don't think there's a risk of it being `None`.

alxtkr77 replied:
Good point. However, there's an edge case: if DSN has a trailing slash with no database name (e.g., `postgres://host:5432/`), `from_dsn` sets `database=""` (empty string). Fixed `from_dsn` to handle empty paths properly, then removed this fallback.

gtopper commented on mlrun/datastore/datastore_profile.py:
```diff
@@ -500,18 +500,37 @@ class DatastoreProfilePostgreSQL(DatastoreProfile):
+
+    def admin_dsn(self) -> str:
+        """
+        Get DSN for administrative operations using the 'postgres' maintenance database.
```
Where does the term "maintenance database" come from? What makes it a maintenance database?

alxtkr77 replied:
"Maintenance database" is [official PostgreSQL terminology](https://www.postgresql.org/docs/current/manage-ag-templatedbs.html) - the `postgres` database always exists and is used for administrative connections. Clarified the docstring.

gtopper replied:
The word "maintenance" doesn't appear in the linked page, though. From what I read, postgres is only the _default_ maintenance database. This can be configured by the database admin so that the maintenance database is something else.

alxtkr77 replied:
You're right - `postgres` is the default and can technically be dropped/replaced. However, it's a reasonable assumption for standard PostgreSQL setups:
- The `postgres` database exists by default in all PostgreSQL installations
- PostgreSQL tools (`createdb`, `pg_dump`, etc.) default to it
- If someone has a non-standard setup without `postgres`, they'd need to provide a custom DSN anyway

Updated the comment and docstring to say "default" instead of implying it's guaranteed.

gtopper commented on mlrun/model_monitoring/db/tsdb/timescaledb/timescaledb_connector.py:
```diff
@@ -71,6 +78,28 @@ def __init__(
+            logger.info(
+                "Auto-generated database name for TimescaleDB",
+                original_database=profile.database,
+                target_database=self.database,
```
Why are we calling it "target database"? Why target?

alxtkr77 replied:
Fair point, renamed to just `database` for clarity.

gtopper replied:
The term "target database" is still used in a lot of other places though.

alxtkr77 replied:
Updated terminology: "target database" → "monitoring database" (more specific to the use case). Also changed "maintenance database" → "default 'postgres' database" for consistency.

gtopper commented on mlrun/model_monitoring/db/tsdb/timescaledb/timescaledb_operations.py:
```diff
@@ -80,6 +83,64 @@ def __init__(
+        Note: Requires a profile to be set during initialization.
+        """
+        if not self._profile:
+            logger.debug(
+                "No profile provided, skipping database creation",
+                project=self.project,
+            )
+            return
+
+        database_name = self._profile.database
+
+        logger.debug(
+            "Checking/creating TimescaleDB database",
```
Should probably be info logs?

alxtkr77 replied:
The "Created TimescaleDB database" log is already at info level. Kept "Checking" at debug to avoid noise on every connector initialization.

gtopper commented on mlrun/model_monitoring/db/tsdb/timescaledb/timescaledb_operations.py:
```diff
@@ -80,6 +83,64 @@ def __init__(
+                project=self.project,
+                database=database_name,
+            )
+        # Note: TimescaleDBConnection uses connection pooling - no explicit close needed
```
Are you sure? `ConnectionPool` does have a close method, and https://www.psycopg.org/psycopg3/docs/advanced/pool.html seems to indicate that it should be called.

alxtkr77 replied:
Good catch. The admin connection (used once for DB creation) should be closed. The main connection is long-lived by design. Added cleanup for the admin connection.

gtopper commented on mlrun/datastore/datastore_profile.py:
```diff
@@ -500,18 +500,37 @@ class DatastoreProfilePostgreSQL(DatastoreProfile):
+        :return: The DSN string.
+        """
+        db = database if database is not None else self.database
+        return f"{self.type}://{self.user}:{self.password}@{self.host}:{self.port}/{db}"
```
Should be fine so long as no one feeds it any special characters like slash or colon.

alxtkr77 replied:
If the user provides a custom database name via DSN, the database must already exist and be valid - no need to validate. The auto-generated name uses `system_id` which is platform-controlled config.

alxtkr77 replied:
Good catch on special characters. Fixed encoding/decoding consistency:
- `from_dsn()` now decodes URL-encoded username/password (urlparse doesn't do this automatically)
- `dsn()` now encodes them back

This ensures round-trip consistency for credentials with special characters like `@`, `:`, `/`. Also fixed the same issue in TDEngine and Redis profiles, and added comprehensive tests.

gtopper commented on mlrun/datastore/datastore_profile.py:
```diff
@@ -500,18 +500,37 @@ class DatastoreProfilePostgreSQL(DatastoreProfile):
+                        If None, uses the configured database.
+        :return: The DSN string.
+        """
+        db = database if database is not None else self.database
```
```suggestion
        db = database or self.database
```

alxtkr77 replied:
Simplified.

gtopper commented on mlrun/datastore/datastore_profile.py:
```diff
@@ -500,18 +500,37 @@ class DatastoreProfilePostgreSQL(DatastoreProfile):
+        """
+        Get DSN for administrative operations using the 'postgres' maintenance database.
+
+        This is useful for operations that need to create/drop databases,
```
useful meaning needed?

alxtkr77 replied:
Already addressed when clarifying the docstring.

gtopper commented on mlrun/datastore/datastore_profile.py:
```diff
@@ -500,18 +500,37 @@ class DatastoreProfilePostgreSQL(DatastoreProfile):
+        Get DSN for administrative operations using the 'postgres' maintenance database.
+
+        This is useful for operations that need to create/drop databases,
+        as you cannot connect to a database while creating it.
```
Probably true, but doesn't seem relevant here?

alxtkr77 replied:
Already addressed when clarifying the docstring.

gtopper commented on tests/model_monitoring/db/tsdb/timescaledb/test_timescaledb_connector.py:
```diff
@@ -0,0 +1,232 @@
+            )
+
+    @contextlib.contextmanager
+    def _config_context(self, system_id=None, auto_create=None):
```
```suggestion
    @staticmethod
    @contextlib.contextmanager
    def _config_context(system_id=None, auto_create=None):
```

alxtkr77 replied:
Added `@staticmethod`.

gtopper commented on tests/model_monitoring/db/tsdb/timescaledb/test_timescaledb_connector.py:
```diff
@@ -0,0 +1,232 @@
+
+        Always restores config even if test fails.
+        """
+        import mlrun
```
Any reason to import mlrun locally? Seems like it belongs at the top.

alxtkr77 replied:
Moved import to top of file.

gtopper commented on tests/model_monitoring/db/tsdb/timescaledb/test_timescaledb_connector.py:
```diff
@@ -0,0 +1,232 @@
+                    with contextlib.suppress(Exception):
+                        connector2.delete_tsdb_resources()
+                # Force drop database
+                self._force_drop_database(admin_connection, expected_db_name)
```
No need to repeat the method name in a comment.
```suggestion
                self._force_drop_database(admin_connection, expected_db_name)
```

alxtkr77 replied:
Removed redundant comments.

gtopper commented on tests/model_monitoring/db/tsdb/timescaledb/test_timescaledb_connector.py:
```diff
@@ -0,0 +1,232 @@
+            dsn=connection_string, profile_name="test_profile"
+        )
+
+    def _force_drop_database(self, admin_connection, database_name):
```
```suggestion
    @staticmethod
    def _force_drop_database(admin_connection, database_name):
```

alxtkr77 replied:
Added `@staticmethod`.

gtopper commented on tests/model_monitoring/db/tsdb/timescaledb/test_timescaledb_connector.py:
```diff
@@ -0,0 +1,232 @@
+        with contextlib.suppress(Exception):
+            admin_connection.run(
+                statements=[f'DROP DATABASE IF EXISTS "{database_name}"']
+            )
```
Might want to consider logging the error, or we'll never know about it.

alxtkr77 replied:
Intentional - the docstring notes "Silently ignores all errors to ensure cleanup doesn't fail tests." For test cleanup, I prefer silent suppression to avoid noisy output. If cleanup consistently fails, we'd notice from leftover databases or subsequent test failures.

gtopper replied:
Subsequent test failures are going to be much harder to tie back to this without any logs, though.

alxtkr77 replied:
Fair point. Changed to log warnings on cleanup failures - this way tests stay clean but debugging info is available when things go wrong.

gtopper commented on tests/model_monitoring/db/tsdb/timescaledb/test_timescaledb_connector.py:
```diff
@@ -0,0 +1,232 @@
+                connector = TimescaleDBConnector(
+                    project=project_name,
+                    profile=profile,
+                )
```
If you move this out of the try, you won't need `connector = None` and `if connector:`.

alxtkr77 replied:
Refactored - moved connector creation outside try block.

gtopper commented on tests/model_monitoring/db/tsdb/timescaledb/test_timescaledb_connector.py:
```diff
@@ -0,0 +1,232 @@
+                    profile=profile,
+                )
+
+            assert "system_id is not set" in str(exc_info.value)
```
Better use `match="..."`.

alxtkr77 replied:
Changed to use `match` parameter.

gtopper reviewed (commented):
LGTM by and large, but had some comments regarding semantics, style, logs, a minor input sanitation concern, and at least one place where we might need to close the connection pool.

Commit:
[Model Monitoring] Address PR review feedback

gtopper commented on mlrun/model_monitoring/db/tsdb/timescaledb/timescaledb_operations.py:
```diff
@@ -80,6 +83,68 @@ def __init__(
+        finally:
+            # Close the admin connection pool to avoid resource leak
+            if admin_connection._pool:
+                admin_connection._pool.close()
```
Should probably create a close method in `TimescaleDBConnection`.

alxtkr77 replied:
Added `close()` method to `TimescaleDBConnection` for proper encapsulation. Updated the admin connection cleanup to use it.

Commit:
[Model Monitoring] Address additional PR review feedback (ML-11662)

Changes based on PR #9012 review:

1. URL encoding consistency in DatastoreProfiles:
   - Add URL encoding for user/password in dsn() methods
   - Add URL decoding in from_dsn() methods (urlparse doesn't decode)
   - Fix affects PostgreSQL, TDEngine, and Redis profiles
   - Add comprehensive tests for URL encoding round-trips

2. Terminology improvements:
   - Replace "maintenance database" with "default postgres database"
   - Replace "target database" with "monitoring database"

3. TimescaleDBConnection improvements:
   - Add close() method for proper encapsulation
   - Update timescaledb_operations to use close() instead of _pool.close()

gtopper commented on tests/datastore/test_datastore_profile.py:
```diff
@@ -175,6 +271,49 @@ def test_from_dsn_without_database() -> None:
+
+    @staticmethod
+    def test_admin_dsn() -> None:
+        """Test admin_dsn() returns DSN pointing to postgres maintenance database."""
```
'maintenance database' here too.

gtopper reviewed (approved):
Other than that, LGTM ✅
